File: WorldSimulatorJAVA/Antilope.java
package worldsimulator;

import java.util.List;
import java.util.Random;

public class Antilope extends Animal {
	
    private static final int RUN_CHANCE = 50;

    public Antilope(World w, Coordinate c) {
        strength = Species.ANTILOPE.getStrength();
        initiative = Species.ANTILOPE.getInitiative();
        species = Species.ANTILOPE.getName();
        sprite = Species.ANTILOPE.getSprite();
        age = 0;
        world = w;
        coord = c;
        dead = false;
        world.addOrganism(this);
    }

    public Antilope(World w, Coordinate c, int s, int a) {
    	strength = s;
    	initiative = Species.ANTILOPE.getInitiative();
        species = Species.ANTILOPE.getName();
        sprite = Species.ANTILOPE.getSprite();
        age = a;
        world = w;
        coord = c;
        dead = false;
        world.addOrganism(this);
    }

    @Override
    public void action() {
    	String message;
        Coordinate oldCoord = new Coordinate(getCoord());
        Coordinate newCoord = new Coordinate(oldCoord);
        List<Coordinate> coordinates = world.getAdjacentTiles(newCoord);
        Random rand = new Random();
        if (coordinates.size() > 0) {
        	int rolledValue = rand.nextInt(coordinates.size());
            newCoord = coordinates.get(rolledValue);
            coordinates = world.getAdjacentTiles(newCoord);
            if (coordinates.size() > 0) {
            	newCoord = oldCoord;
            	while (newCoord.equals(oldCoord)) {
            	rolledValue = rand.nextInt(coordinates.size());
                newCoord = coordinates.get(rolledValue);
            	}
            }
        }

        Organism opponent = world.findOrganismByCoordinate(newCoord);
        message = String.format(getSpecies() + getID() + " przechodzi z pola " + oldCoord + " na pole " + newCoord);
        world.prompt(message);
        if (opponent == null) {
        	message = String.format("\n");
        	world.prompt(message);
            setCoord(newCoord);
        } else {
            collision(this, opponent);
        }
    }

    @Override
	public boolean ranAway(Organism attacker) {
    	String message;
        Random rand = new Random();
        int runRoll = rand.nextInt(100);
        if (runRoll < RUN_CHANCE) {
            Coordinate newCoord = findVoidSpace(getCoord());
            if (!newCoord.equals(getCoord())) {
                attacker.setCoord(getCoord());
                message = String.format(getSpecies() + getID() + " uciekla przed " + attacker.getSpecies() + attacker.getID() + " z pola " + getCoord() + " na pole " + newCoord + "\n");
                world.prompt(message);
                setCoord(newCoord);
                return true;
            } else {
            	message = String.format(getSpecies() + getID() + "Nie moze uciec - brak pustego miejsca w poblizu\n");
                world.prompt(message);
            	return false;
            }
        } else {
            return false;
        }
    }
}

File: WorldSimulatorCPP/Antilope.cpp
#include "Antilope.h"
#include "World.h"
Antilope::Antilope(World* w, Coordinate c) {
    strength = 4;
    initiative = 4;
    species = "Antilope";
    sprite = 'A';
    age = 0;
    world = w;
    coord = c;
    dead = false;
    w->addOrganism(this);
}

Antilope::Antilope(World* w, Coordinate c, int s, int a) {
    strength = s;
    initiative = 4;
    species = "Antilope";
    sprite = 'A';
    age = a;
    world = w;
    coord = c;
    dead = false;
    w->addOrganism(this);
}

void Antilope::action() {
    Coordinate oldCoord = getCoord();
    Coordinate newCoord = oldCoord;
    while (newCoord == oldCoord)
    {
        int dir = (rand() % 4);
        if (dir == North) {
            newCoord.y -= 2;
            if (newCoord.y < 0) {
                newCoord.y += 2;
            }
        }
        else if (dir == East) {
            newCoord.x += 2;
            if (newCoord.x >= world->getColumns()) {
                newCoord.x -= 2;
            }
        }
        else if (dir == South) {
            newCoord.y += 2;
            if (newCoord.y >= world->getRows()) {
                newCoord.y -= 2;
            }
        }
        else if (dir == West) {
            newCoord.x -= 2;
            if (newCoord.x < 0) {
                newCoord.x += 2;
            }
        }
    }
    Organism* opponent = world->findOrganismByCoordinate(newCoord);
    cout << getSpecies() << getID() << " przechodzi z pola " << oldCoord << " na pole " << newCoord << endl;
    if (opponent == nullptr)
    {
        cout << endl;
        setCoord(newCoord);
    }
    else
        collision(this, opponent);
}

bool Antilope::ranAway(Organism* attacker) {
    int runRoll = rand() % 100;
    if (runRoll < RUN_CHANCE) {
        Coordinate newCoord = FindVoidSpace(getCoord()); // wybranie miejsca ucieczki
        if (newCoord != getCoord())
        {
            attacker->setCoord(getCoord());
            cout << getSpecies() << getID() << " uciekla przed " << attacker->getSpecies() << attacker->getID() << " z pola " << getCoord() << " na pole " << newCoord << endl << endl;
            setCoord(newCoord);
            return true;
        }
        else {
            cout << getSpecies() << getID() << "Nie moze uciec - brak pustego miejsca w poblizu" << endl;
            return false;
        }
    }
    else return false;
}

void Antilope::draw() const {
    cout << YELLOW << getSprite() << RESET;
}